### DISCUSSION TITLE
[GeanyLua Plugin]  Add specified number of blank lines

### DISCUSSION THREAD

**Discussion by advice2020** [geany/geany#4091] [General]
GeanyLua Plugin

Was wondering if anyone would be willing to provide two Lua scripts that when applied, adds a specified number of blank lines to current document.  Basically the equivalent of hitting the "Enter" key over and over “x” amount of times.

1.) Add 15 blank lines below line that cursor is currently in

2.) Add 25 blank lines to the bottom of the current document regardless of where cursor is.

Thank You

**Comment by Skif-off** [comment]
Maybe something like this
```lua
local function GetEmptyLines(sLE, iLE)
  local sR = ""
  for i = 1, iLE do sR = sR .. sLE end
  return sR
end

local aEOL = {"\r\n", "\r", "\n"}
local iEOL = geany.scintilla("SCI_GETEOLMODE")
local sAdd = GetEmptyLines(aEOL[iEOL + 1], 15)

local iCurPos = geany.caret()
local iL, iC = geany.rowcol(iCurPos)
iCurPos = geany.scintilla("SCI_GETLINEENDPOSITION", iL - 1)
geany.caret(iCurPos)
geany.selection(sAdd)
```
```lua
local function GetEmptyLines(sLE, iLE)
  local sR = ""
  for i = 1, iLE do sR = sR .. sLE end
  return sR
end

local aEOL = {"\r\n", "\r", "\n"}
local iEOL = geany.scintilla("SCI_GETEOLMODE")
local sAdd = GetEmptyLines(aEOL[iEOL + 1], 25)

local sTxt = geany.text()
geany.text(sTxt .. sAdd)
```
?

**Comment by advice2020** [comment]
@Skif-off
Thank you for providing these tools.
I was able to test them out and here are the results.

-----------------------------------------------------

`ADD LINES BELOW CURSOR TOOL`
Cursor
Would it be possible to have the cursor show up at the top line of those new rows added?
So for example if this tool was applied while cursor is in line 1, it adds 15 lines so now document shows 16 total lines and the cursor is at the beginning of line 2.
Current behavior is cursor ends up in line 16.
Which now when I think about this I actually have a use for this current behavior as well.

-----------------------------------------------------

`ADD LINES TO BOTTOM OF DOCUMENT TOOL`
Cursor
With this tool the cursor ends up moving automatically to the top of the document after adding the new lines to the bottom.  
EDIT:
Would it be possible to have the cursor show up in the last line at the bottom after all the blank lines have been added?

-----------------------------------------------------

I tried looking this up in GeanyLua, strange that there are mentionings of "cursor" when searching the page.
I ended up finding "geany.rowcol" but I would not have known this if I didn't remember seeing this in your script.
Would think the word "cursor" would be used at least once to describe something in the GeanyLua documentation :)

Thank you again for all your help

**Comment by Skif-off** [comment]
In theory:
```lua
local function GetEmptyLines(sLE, iLE)
  local sR = ""
  for i = 1, iLE do sR = sR .. sLE end
  return sR
end

local aEOL = {"\r\n", "\r", "\n"}
local iEOL = geany.scintilla("SCI_GETEOLMODE")
local sAdd = GetEmptyLines(aEOL[iEOL + 1], 15)

local iCurPos = geany.caret()
local iL, iC = geany.rowcol(iCurPos)
iCurPos = geany.scintilla("SCI_GETLINEENDPOSITION", iL - 1)
geany.caret(iCurPos)
geany.selection(sAdd)
geany.caret(iCurPos)
```

```lua
local function GetEmptyLines(sLE, iLE)
  local sR = ""
  for i = 1, iLE do sR = sR .. sLE end
  return sR
end

local aEOL = {"\r\n", "\r", "\n"}
local iEOL = geany.scintilla("SCI_GETEOLMODE")
local sAdd = GetEmptyLines(aEOL[iEOL + 1], 25)

local iLines = geany.height()
local iCurPos = geany.rowcol(iLines + 1, 0)
local sTxt = geany.text()
geany.text(sTxt .. sAdd)
geany.caret(iCurPos)
```

> Adding Precision

Try this version of `GetEmptyLines`, only for tests
```lua
local function GetEmptyLines(sLE, iLE)
  local sR = ""
  for i = 1, iLE do sR = sR .. "::" .. i .. "::" .. sLE end
  return sR
end
```

> Would think the word "cursor" would be used at least once to describe something in the GeanyLua documentation :)

Try "caret" :)

**Comment by advice2020** [comment]
@Skif-off
Thank you for working on these tools.

`PRECISION` (I will remove all of this once I know you have read it)
Ignore everything I mentioned about precision I think everything works fine.
I have removed/cleaned up my posts above about this already.
If feel like it, you can remove the precision code you recently mentioned above to eliminate any confusion for future readers of this post.
Turns out two things occurred in my tests that confused/generated that off by 1 precision.
One of them might be a bug that I will post about.

`CARET TERMINOLOGY`
I have tried them before, they add a nice splash of orange to any salad :)
I completely overlooked that prefix, I was focusing on `iCurPos` which I figured meant something like Cursor Position, but you would think the word cursor would be used at least once somewhere in the documentation :)
I do not think I ever heard it called a “caret” before.
This is the symbol I think of when I hear the word caret “^” (I put the symbol in quotes, it is not meant to be a smug bird sticking its beak up at you :)

`CURSOR/CARET`
These tools add the blank lines which work great, now I think they are all about cursor/caret position.  The tools you have provided offered two different cursor positions then what I was looking for, which worked out because I ended up liking them both :) but this now brings this up to four different versions (just different caret locations).  I will provide four categories to hopefully clearify, if still interested in working on this.

Thank You

**Comment by advice2020** [comment]
`TOOL 1`
`ADD LINES BELOW CURSOR - Cursor at Top`
With the new changes you made, you are very close.
This recent version moves cursor to end of the current line. 
Example: Add 5 Lines
I = cursor/caret

5 Iapple    (run script from this line)
6 banana

5 appleI      (moves cursor to the end of the line script is run in)
6 <---------------(looking to have it show at the beginning of the next line here)
7
8
9
10
11 banana

--------------------------------------------------------------------------

`TOOL 2`
`ADD LINES BELOW CURSOR - Cursor at Bottom`
This one is done already, provided with original script you created for this.

--------------------------------------------------------------------------

`TOOL 3`
`ADD LINES TO BOTTOM OF DOCUMENT - Cursor at Top`
The recent version provided puts the cursor two lines above the first new blank line added

So if document had 10 lines total
7
8
9 I (currently puts cursor here)
10
11 (start of new added lines where would like cursor to start)

--------------------------------------------------------------------------

`TOOL 4`
`ADD LINES TO BOTTOM OF DOCUMENT - Cursor at Bottom`
I actually was able to modify TOOL 3 by adding some lines from TOOL 2 and was able to get this to work, but it ends up doing the same as TOOL 3, where it puts the cursor two lines above the very bottom blank line added.

--------------------------------------------------------------------------

Hopefully all of this makes sense.
If still interested, again all of this is really just....
-cursor at top of new blank lines added 
and 
-cursor at bottom of new blank lines added

Thank you again for working on this

**Comment by Skif-off** [comment]
In theory:
```lua
local function GetEmptyLines(sLE, iLE)
  local sR = ""
  for i = 1, iLE do sR = sR .. sLE end
  return sR
end

local aEOL = {"\r\n", "\r", "\n"}
local iEOL = geany.scintilla("SCI_GETEOLMODE")
local sAdd = GetEmptyLines(aEOL[iEOL + 1], 15)

local iCurPos = geany.caret()
local iL, iC = geany.rowcol(iCurPos)
iCurPos = geany.scintilla("SCI_GETLINEENDPOSITION", iL - 1)
geany.caret(iCurPos)
geany.selection(sAdd)
geany.caret(iCurPos + string.len(aEOL[iEOL + 1]))
```

```lua
local function GetEmptyLines(sLE, iLE)
  local sR = ""
  for i = 1, iLE do sR = sR .. sLE end
  return sR
end

local aEOL = {"\r\n", "\r", "\n"}
local iEOL = geany.scintilla("SCI_GETEOLMODE")
local sAdd = GetEmptyLines(aEOL[iEOL + 1], 25)

local iLines = geany.height()
local sTxt = geany.text()
geany.text(sTxt .. sAdd)
local iCurPos = geany.rowcol(iLines + 1, 0)
geany.caret(iCurPos)
```

It seems, the second script did not work correctly:
```lua
local iCurPos = geany.rowcol(iLines + 1, 0)
```
but in this moment, the line with the number `iLines + 1` does not yet exist.

P.S.

> I do not think I ever heard it called a “caret” before.

"Cursor" can be either a "text cursor" or a "mouse cursor", maybe this is the reason? In any case, this word is an adequate word and can be seen in the documentation of various projects.

**Comment by advice2020** [comment]
@Skif-off
Thank you for continuing to work on these tools.
The two new scripts you provided work great.
The only one I had a question on was TOOL 4

`TOOL 1`
`ADD LINES BELOW CURSOR - Cursor at Top`
done

`TOOL 2`
`ADD LINES BELOW CURSOR - Cursor at Bottom`
done

`TOOL 3`
`ADD LINES TO BOTTOM OF DOCUMENT - Cursor at Top`
done

`TOOL 4`
`ADD LINES TO BOTTOM OF DOCUMENT - Cursor at Bottom`
Do you mean that this one cannot be achieved?

I had mentioned I tried combining your TOOL 2 with older version of TOOL 3 and it kind of works.  The first time I tested this the document I used for the test had blank lines at the end.  I just tested it again on a document that did have text in the last lines and realized that it actually appears that it is more like adding new blank lines in the line number right before the end, because that last line (with or without text in it) actually ends up at the bottom of the document.
Again when last line is already blank it gives illusion that it is almost working.

Here is that code
```lua
local function GetEmptyLines(sLE, iLE)
  local sR = ""
  for i = 1, iLE do sR = sR .. sLE end
  return sR
end

local aEOL = {"\r\n", "\r", "\n"}
local iEOL = geany.scintilla("SCI_GETEOLMODE")
local sAdd = GetEmptyLines(aEOL[iEOL + 1], 10)

local iLines = geany.height()
local iCurPos = geany.caret()
local iL, iC = geany.rowcol(iCurPos)
iCurPos = geany.rowcol(iLines + 1, 0)
geany.caret(iCurPos)
geany.selection(sAdd)
```
Probably some issues in it, I just added some random lines to the bottom to see what would happen :)

I was wondering, to achieve TOOL 4, would it be possible to add a line to the latest version you just provided for TOOL 3, where after the new blank lines are added, then a new instruction to just move the caret to the bottom line of the document?  Only thing I can think of.

Thank you again for your help with this.

**Comment by Skif-off** [comment]
_ADD LINES TO BOTTOM OF DOCUMENT - Cursor at Bottom_
For example:
```lua
local function GetEmptyLines(sLE, iLE)
  local sR = ""
  for i = 1, iLE do sR = sR .. sLE end
  return sR
end

local aEOL = {"\r\n", "\r", "\n"}
local iEOL = geany.scintilla("SCI_GETEOLMODE")
local sAdd = GetEmptyLines(aEOL[iEOL + 1], 25)

local sTxt = geany.text()
geany.text(sTxt .. sAdd)
geany.select(geany.length())
```

**Comment by advice2020** [comment]
@Skif-off
Ahh, thank you for working on that one as well, works great.
Okay so now the cursor location works in all four tools.
I noticed something but I am not sure if it is a Geany based issue or a GeanyLua issue, but maybe can be corrected within the scripts for the following two tools, because same issue occurs in both of them.

`TOOL 2`
`ADD LINES BELOW CURSOR - Cursor at Bottom`
AND
`TOOL 4`
`ADD LINES TO BOTTOM OF DOCUMENT - Cursor at Bottom`

When using either of the above tools, if new blank lines are added that end up beyond visible line numbers shown in document/Geany window, I would like Geany to scroll down to that new cursor location automatically so that line/cursor location is visible on screen.
Strange because sometimes it does this and sometimes it does not.
Not sure why this occurs?
If nothing stands out to you why, would you be able to add code that automatically performs this scroll to line/cursor action each time either of these scripts (TOOL2 & TOOL4) are run, to override whatever is causing this.

Thank You

**Comment by Skif-off** [comment]
I found [SCI_SCROLLCARET](https://scintilla.org/ScintillaDoc.html#SCI_SCROLLCARET), try to add `geany.scintilla("SCI_SCROLLCARET")` to the end.

Also I found [SCI_GOTOLINE](https://scintilla.org/ScintillaDoc.html#SCI_GOTOLINE), this way will require more changes. For example, [second script](https://github.com/geany/geany/discussions/4091#discussioncomment-11628236):
try to replace
```lua
local iLines = geany.height()
local sTxt = geany.text()
geany.text(sTxt .. sAdd)
local iCurPos = geany.rowcol(iLines + 1, 0)
geany.caret(iCurPos)
```
on
```lua
local iLines = geany.height()
local sTxt = geany.text()
geany.text(sTxt .. sAdd)
geany.scintilla("SCI_GOTOLINE", iLines + 1)
```

**Comment by advice2020** [comment]
@Skif-off
Oh I was looking in GeanyLua documentation, good idea I was not thinking of searching scintilla.  
Speaking of which, wow, you got it with the first one.
Adding the `geany.scintilla("SCI_SCROLLCARET")` you mentioned to the bottom of TOOL2 & TOOL4 works perfectly!
Believe it or not I think this is the last GeanyLua request that I have at the moment :)
You have been so helpful, thank you so much for all of these great tools you have provided, as well as for answering the questions I have asked.  Hopefully our paths will cross again in the future :)
Thank you my friend
